dashboard/app/models/course.rb

```diff
@@ -167,20 +167,25 @@ def update_scripts(new_scripts, alternate_scripts = nil)
 
   # Get the assignable info for this course, then update translations
   # @return AssignableInfo
-  def assignable_info
+  def assignable_info(user = nil)
     info = ScriptConstants.assignable_info(self)
     # ScriptConstants gives us untranslated versions of our course name, and the
     # category it's in. Set translated strings here
     info[:name] = localized_title
     info[:category] = I18n.t('courses_category')
-    info[:script_ids] = default_course_scripts.map(&:script_id)
+    info[:script_ids] = user ?
+      scripts_for_user(user).map(&:id) :
+      default_course_scripts.map(&:script_id)
     info
   end
 
-  # Get the set of valid courses for the dropdown in our sections table. This should
-  # be static data, but contains localized strings so we can only cache on a per
-  # locale basis
-  def self.valid_courses
+  # Get the set of valid courses for the dropdown in our sections table. This
+  # should be static data for users without experiments enabled, but contains
+  # localized strings so we can only cache on a per locale basis.
+  def self.valid_courses(user = nil)
+    # Do not cache if the user might have an experiment enabled which puts them
+    # on an alternate script.
+    return Course.courses_for_user_with_experiments(user) if has_any_course_experiments?(user)
```

Comment:
Should this be `if user && has_any_course_experiments?(user)` to avoid an experiment lookup when we have no user? It looks like that look up is cached anyways, so it may not really matter.